File: apps/android/core/src/main/kotlin/diy/atelier/core/OfflineStatus.kt
package diy.atelier.core

import java.time.Instant

enum class PersistencePhase {
    LOCAL_ONLY,
    QUEUED,
    SYNCING,
    DURABLE,
    NEEDS_ATTENTION,
}

data class OfflineStatus(
    val phase: PersistencePhase,
    val pendingChangeCount: Int = 0,
    val lastDurableAt: Instant? = null,
) {
    init {
        require(pendingChangeCount >= 0)
    }

    companion object {
        val LocalOnly = OfflineStatus(PersistencePhase.LOCAL_ONLY)
    }
}

File: apps/apple/Packages/AtelierShared/Sources/AtelierCore/OfflineStatus.swift
import Foundation

public enum PersistencePhase: String, Codable, CaseIterable, Sendable {
    case localOnly
    case queued
    case syncing
    case durable
    case needsAttention
}

public struct OfflineStatus: Codable, Equatable, Sendable {
    public var phase: PersistencePhase
    public var pendingChangeCount: Int
    public var lastDurableAt: Date?

    public init(
        phase: PersistencePhase,
        pendingChangeCount: Int = 0,
        lastDurableAt: Date? = nil
    ) {
        self.phase = phase
        self.pendingChangeCount = max(0, pendingChangeCount)
        self.lastDurableAt = lastDurableAt
    }

    public static let localOnly = OfflineStatus(phase: .localOnly)
}
